src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.net7.Manifest/WorkloadManifest.targets.in

```diff
@@ -103,8 +103,8 @@
       <KnownRuntimePack Update="@(KnownRuntimePack)">
         <LatestRuntimeFrameworkVersion Condition="'%(KnownRuntimePack.TargetFramework)' == 'net7.0' and '%(KnownRuntimePack.RuntimePackLabels)' == 'Mono'">$(_MonoWorkloadRuntimePackPackageVersion)</LatestRuntimeFrameworkVersion>
         <!-- Overrides for wasm threading support -->
-        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreading)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.net7.**RID**</RuntimePackNamePatterns>
-        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnablePerfTrace)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.perftrace.net7.**RID**</RuntimePackNamePatterns>
+        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreads)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.net7.**RID**</RuntimePackNamePatterns>
```

Comment:
@lewing @radical @steveisok So actually this change doesn't work.  I got nuget errors complaining that it couldn't find `Microsoft.NETCore.App.Runtime.Mono.multithread.net7.browser-wasm` in the `dotnet7` feed.

What does work is
```suggestion
        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreads)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.**RID**</RuntimePackNamePatterns>
```

But that feels wrong.  shouldn't we be publishing the variant runtime packs with `net7` in the name?

Reply:
Or maybe this needs to be changed after all.  If I look at https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet7, I only see packages like `Microsoft.NETCore.App.Runtime.Mono.multithread.browser-wasm` and `Microsoft.NETCore.App.Runtime.Mono`.  The only things with a `net7` component to the name are manifests.

Reply:
Mapping:

https://github.com/lambdageek/runtime/blob/85789bd8e4b01c3c85b760a686c7f66c294a26c9/src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.net7.Manifest/WorkloadManifest.json.in#L441-L446

Reply:
@radical I see. so the manifest is supposed to take care of the mapping. Is there some reason why the `artifacts/bin/dotnet-net7` SDK (from WBT) isn't able to find the `Microsoft.NETCore.App.Runtime.Mono.multithread.net7.browser-wasm` package? does it not use the mapping?

Reply:
What is failing exactly? What is trying to find this package?

Reply:
@radical after creating a `dotnet-net7` SDK, I then use it to do `dotnet new wasmbrowser` and then do a threaded build:

```console
$ dotnet build -bl -p:WasmEnableThreads=true -p:WasmGenerateAppBundle=true
MSBuild version 17.4.0-preview-22416-02+5d102ae37 for .NET
/home/aleksey/working/dotnet-runtime/artifacts/bin/dotnet-net7/sdk/7.0.100-rc.2.22456.6/MSBuild.dll --property:WasmEnableThreads=true --property:WasmGenerateAppBundle=true -bl -consoleloggerparameters:Summary -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/home/aleksey/working/dotnet-runtime/artifacts/bin/dotnet-net7/sdk/7.0.100-rc.2.22456.6/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/home/aleksey/working/dotnet-runtime/artifacts/bin/dotnet-net7/sdk/7.0.100-rc.2.22456.6/dotnet.dll -maxcpucount -restore -verbosity:m ./boba.csproj
  Determining projects to restore...
/home/aleksey/hacking/net7-playground/boba/boba.csproj : error NU1101: Unable to find package Microsoft.NETCore.App.Runtime.Mono.multithread.net7.browser-wasm. No packages exist with this id in source(s): dotnet7, nuget.org
  Failed to restore /home/aleksey/hacking/net7-playground/boba/boba.csproj (in 484 ms).

Build FAILED.

/home/aleksey/hacking/net7-playground/boba/boba.csproj : error NU1101: Unable to find package Microsoft.NETCore.App.Runtime.Mono.multithread.net7.browser-wasm. No packages exist with this id in source(s): dotnet7, nuget.org
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.59

Workload updates are available. Run `dotnet workload list` for more information.
```

from the binlog, it looks like the Restore task is failing.

[msbuild.binlog.zip](https://github.com/dotnet/runtime/files/9502333/msbuild.binlog.zip)

The `_GenerateRestoreGraph` target invokes msbuild again on the project and runs `ProcessFrameworkReferences` which adds a `RuntimePack` item with identity `Microsoft.NETCore.App.Runtime.Mono.multithread.net7.browser-wasm`

Reply:
It should mention the version too. Try adding the shipping packages directory in artifacts as a nuget source.

Reply:
didn't seem to make a difference.  (I did `nuget init .../artifacts/packages/Release/Shipping .../packages`), then used it as another source in NuGet.config.  this is what is in that local feed:

```
$ tree ../packages/
../packages/
├── microsoft.netcore.app.ref
│   └── 7.0.0-dev
│       ├── microsoft.netcore.app.ref.7.0.0-dev.nupkg
│       ├── microsoft.netcore.app.ref.7.0.0-dev.nupkg.sha512
│       └── microsoft.netcore.app.ref.nuspec
├── microsoft.netcore.app.runtime.aot.linux-x64.cross.browser-wasm
│   └── 7.0.0-dev
│       ├── microsoft.netcore.app.runtime.aot.linux-x64.cross.browser-wasm.7.0.0-dev.nupkg
│       ├── microsoft.netcore.app.runtime.aot.linux-x64.cross.browser-wasm.7.0.0-dev.nupkg.sha512
│       └── microsoft.netcore.app.runtime.aot.linux-x64.cross.browser-wasm.nuspec
├── microsoft.netcore.app.runtime.mono.browser-wasm
│   └── 7.0.0-dev
│       ├── microsoft.netcore.app.runtime.mono.browser-wasm.7.0.0-dev.nupkg
│       ├── microsoft.netcore.app.runtime.mono.browser-wasm.7.0.0-dev.nupkg.sha512
│       └── microsoft.netcore.app.runtime.mono.browser-wasm.nuspec
├── microsoft.netcore.app.runtime.mono.multithread.browser-wasm
│   └── 7.0.0-dev
│       ├── microsoft.netcore.app.runtime.mono.multithread.browser-wasm.7.0.0-dev.nupkg
│       ├── microsoft.netcore.app.runtime.mono.multithread.browser-wasm.7.0.0-dev.nupkg.sha512
│       └── microsoft.netcore.app.runtime.mono.multithread.browser-wasm.nuspec
├── microsoft.netcore.app.runtime.mono.perftrace.browser-wasm
│   └── 7.0.0-dev
│       ├── microsoft.netcore.app.runtime.mono.perftrace.browser-wasm.7.0.0-dev.nupkg
│       ├── microsoft.netcore.app.runtime.mono.perftrace.browser-wasm.7.0.0-dev.nupkg.sha512
│       └── microsoft.netcore.app.runtime.mono.perftrace.browser-wasm.nuspec
├── microsoft.net.runtime.monoaotcompiler.task
│   └── 7.0.0-dev
│       ├── microsoft.net.runtime.monoaotcompiler.task.7.0.0-dev.nupkg
│       ├── microsoft.net.runtime.monoaotcompiler.task.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.runtime.monoaotcompiler.task.nuspec
├── microsoft.net.runtime.monotargets.sdk
│   └── 7.0.0-dev
│       ├── microsoft.net.runtime.monotargets.sdk.7.0.0-dev.nupkg
│       ├── microsoft.net.runtime.monotargets.sdk.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.runtime.monotargets.sdk.nuspec
├── microsoft.net.runtime.wasm.sample.mono
│   └── 7.0.0-dev
│       ├── microsoft.net.runtime.wasm.sample.mono.7.0.0-dev.nupkg
│       ├── microsoft.net.runtime.wasm.sample.mono.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.runtime.wasm.sample.mono.nuspec
├── microsoft.net.runtime.webassembly.sdk
│   └── 7.0.0-dev
│       ├── microsoft.net.runtime.webassembly.sdk.7.0.0-dev.nupkg
│       ├── microsoft.net.runtime.webassembly.sdk.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.runtime.webassembly.sdk.nuspec
├── microsoft.net.runtime.webassembly.templates
│   └── 7.0.0-dev
│       ├── microsoft.net.runtime.webassembly.templates.7.0.0-dev.nupkg
│       ├── microsoft.net.runtime.webassembly.templates.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.runtime.webassembly.templates.nuspec
├── microsoft.net.workload.mono.toolchain.net6.manifest-7.0.100
│   └── 7.0.0-dev
│       ├── microsoft.net.workload.mono.toolchain.net6.manifest-7.0.100.7.0.0-dev.nupkg
│       ├── microsoft.net.workload.mono.toolchain.net6.manifest-7.0.100.7.0.0-dev.nupkg.sha512
│       └── microsoft.net.workload.mono.toolchain.net6.manifest-7.0.100.nuspec
└── microsoft.net.workload.mono.toolchain.net7.manifest-7.0.100
    └── 7.0.0-dev
        ├── microsoft.net.workload.mono.toolchain.net7.manifest-7.0.100.7.0.0-dev.nupkg
        ├── microsoft.net.workload.mono.toolchain.net7.manifest-7.0.100.7.0.0-dev.nupkg.sha512
        └── microsoft.net.workload.mono.toolchain.net7.manifest-7.0.100.nuspec

24 directories, 36 files
```

Reply:
I will have to try this locally tomorrow.
Not sure what `dotnet nuget init` does. I usually just add the file:/// path to my nuget.config

Reply:
> I will have to try this locally tomorrow. Not sure what `dotnet nuget init` does. I usually just add the file:/// path to my nuget.config

`nuget init` just copies the nugets and makes the directory layout.

In any case, I tried it with just adding the artifacts shipping dir as a `file:///` source and that didn't work either. same error.

Reply:
@radical @lewing I'm pretty sure it's supposed to be without the `.net7` component. 
I added this to the WorkloadManifest.targets

```xml
        <!-- Overrides for wasm threading support -->
	<_AlekseyPreviousPatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreads)' == 'true'">%(KnownRuntimePack.RuntimePackNamePatterns)</_AlekseyPreviousPatterns>
        <RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreads)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.net7.**RID**</RuntimePackNamePatterns>
```

and then in the msbuild log for my sample project I can see

```
Microsoft.NETCore.App
    TargetFramework = net7.0
    LatestRuntimeFrameworkVersion = 7.0.0-dev
    RuntimePackRuntimeIdentifiers = linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;win-arm;win-arm64;win-x64;win-x86;linux-musl-arm;osx-arm64;maccatalyst-x64;maccatalyst-arm64;linux-s390x;linux-bionic-arm;linux-bionic-arm64;linux-bionic-x64;linux-bionic-x86;browser-wasm;ios-arm64;ios-arm;iossimulator-arm64;iossimulator-x64;iossimulator-x86;tvos-arm64;tvossimulator-arm64;tvossimulator-x64;android-arm64;android-arm;android-x64;android-x86
    RuntimePackNamePatterns = Microsoft.NETCore.App.Runtime.Mono.multithread.net7.**RID**
    _AlekseyPreviousPatterns = Microsoft.NETCore.App.Runtime.Mono.**RID**
    RuntimePackLabels = Mono
    RuntimeFrameworkName = Microsoft.NETCore.App
```

So the workload `alias-to` stuff has already been processed by the time the `KnownRuntimePack` update happens

Reply:
```xml
<RuntimePackNamePatterns Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(WasmEnableThreads)' == 'true'">Microsoft.NETCore.App.Runtime.Mono.multithread.**RID**</RuntimePackNamePatterns>
```

Like you said, this change does work. The `alias-to` stuff would get used by the workload resolver, and the rest would be expecting the actual package names.

With this change I can build the project with `dotnet build -bl -p:WasmEnableThreads=true -p:WasmGenerateAppBundle=true` using `dotnet-net7`.

Reply:
Including `net7` and `net6` in the workload pack names is needed to make them unique inside the manifest.  As ankit said, `alias-to` allows us to map back to how the runtime packs are already named and organized by version.

Reply:
> Including `net7` and `net6` in the workload pack names is needed to make them unique inside the manifest. As ankit said, `alias-to` allows us to map back to how the runtime packs are already named and organized by version.

Yes, I understand that.  The point is by the time this .targets file is doing its thing, the aliases have already done their job.  so when we update the `KnownRuntimePack` the names are already real .nupkg names, not aliases.  so the `.net7.` is not necessary in this .targets